### PR TITLE
WT-7758 Force evict a page when the update chain is too long.

### PR DESCRIPTION
**Pull request by raviprakashgiri29**
This ticket is created as part of BF-21313, to fix the 156% increase in write latency.

### CODE REVIEW

**Comment by quchenhao** [comment]
@raviprakashgiri29 Why you choose the 1000 magic number?

**Comment by raviprakashgiri29** [comment]
> @raviprakashgiri29 Why you choose the 1000 magic number?

@quchenhao It is the average number of update count which I found from my MixedWrites.yml workload stats. I am yet to discuss this with @agorrod .

File: dist/stat_data.py [hunk 1]
```diff
@@ -213,6 +213,7 @@ def __init__(self, name, desc, flags=''):
     CacheStat('cache_eviction_fail_in_reconciliation', 'pages selected for eviction unable to be evicted because of failure in reconciliation'),
     CacheStat('cache_eviction_fail_parent_has_overflow_items', 'pages selected for eviction unable to be evicted as the parent page has overflow items'),
     CacheStat('cache_eviction_force', 'forced eviction - pages selected count'),
+    CacheStat('cache_eviction_force_update_list', 'forced eviction when more than 1000 updates found in an update chain.'),
```
Comment (sulabhM) [inline dist/stat_data.py:216]: Not fond of the statistics naming, maybe we can call it:

`cache_eviction_force_long_update_list`

Also, we should probably not mention the exact number of updates we force the page for eviction. If we tune that value tomorrow, we will have to change the description of the statistics, which is hard-coupled to t2/server-status output. So I suggest something like:

`forced eviction - pages selected because of a large number of updates to a single item`

File: src/btree/row_modify.c [hunk 2]
```diff
@@ -311,13 +311,13 @@ __wt_row_insert_alloc(WT_SESSION_IMPL *session, const WT_ITEM *key, u_int skipde
...
 __wt_update_obsolete_check(
-  WT_SESSION_IMPL *session, WT_PAGE *page, WT_UPDATE *upd, bool update_accounting)
+  WT_SESSION_IMPL *session, WT_PAGE *page, WT_UPDATE *upd, bool update_accounting, u_int *upd_count)
 {
     WT_TXN_GLOBAL *txn_global;
     WT_UPDATE *first, *next;
     size_t size;
     uint64_t oldest, stable;
-    u_int count, upd_seen, upd_unstable;
+    u_int upd_seen, upd_unstable;
```
Comment (sulabhM) [inline src/btree/row_modify.c:321]: We should initialize upd_count to 0 here as a side effect of this function call.

File: src/include/serial_inline.h [hunk 3]
```diff
@@ -224,11 +224,14 @@ __wt_update_serial(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_PAGE *page
     WT_UPDATE *obsolete, *upd;
     wt_timestamp_t obsolete_timestamp, prev_upd_ts;
     uint64_t txn;
+    u_int upd_count;
 
     /* Clear references to memory we now own and must free on error. */
     upd = *updp;
     *updp = NULL;
     prev_upd_ts = WT_TS_NONE;
+    upd_count = 0;
```
Comment (sulabhM) [inline src/include/serial_inline.h:233]: We do not need to initialize upd_count here, I will prefer this be done as a side effect of calling __wt_update_obsolete_check

File: src/include/serial_inline.h [hunk 4]
```diff
@@ -296,11 +299,16 @@ __wt_update_serial(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_PAGE *page
...
         return (0);
 
-    obsolete = __wt_update_obsolete_check(session, page, upd->next, true);
+    obsolete = __wt_update_obsolete_check(session, page, upd->next, true, &upd_count);
 
     WT_PAGE_UNLOCK(session, page);
 
     __wt_free_update_list(session, &obsolete);
 
+    if (upd_count > 1000) {
```
Comment (sulabhM) [inline src/include/serial_inline.h:308]: Instead of 1000 we can say WT_THOUSAND

File: src/include/serial_inline.h [hunk 5]
```diff
@@ -296,11 +299,16 @@ __wt_update_serial(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_PAGE *page
...
+        WT_STAT_CONN_INCR(session, cache_eviction_force_update_list);
+        __wt_page_evict_soon(session, cbt->ref);
```
Comment (sulabhM) [inline src/include/serial_inline.h:310]: Are we guaranteed that cbt->ref will not be NULL. I would think so, this being a serial_update function, but @agorrod can you confirm ?

**Review by sulabhM — CHANGES_REQUESTED** [review]

File: src/btree/row_modify.c [hunk 6]
```diff
@@ -334,7 +334,7 @@ __wt_update_obsolete_check(
      * Only updates with globally visible, self-contained data can terminate update chains.
      *
      */
-    for (first = NULL, count = 0; upd != NULL; upd = upd->next, count++) {
+    for (first = NULL; upd != NULL; upd = upd->next, (*upd_count)++) {
```
Comment (sulabhM) [inline src/btree/row_modify.c:337]: We should try to protect __wt_update_obsolete_check if called with a NULL upd_count. It should be able to function if a NULL is passed. We can keep and use the local "count" variable, and if the caller has passed a non NULL "upd_count", we can populate it with the update count. This sort of usage is common in the WiredTiger code.

As an example look at __wt_meta_ckptlist_get_from_config(). It takes an optional "size_t *allocatedp". It uses a local variable, "allocated" for the needed functionality, but at the end copies the result into "allocatedp" if not NULL.

File: src/btree/row_modify.c [hunk 7]
```diff
@@ -311,13 +311,13 @@ __wt_row_insert_alloc(WT_SESSION_IMPL *session, const WT_ITEM *key, u_int skipde
  */
 WT_UPDATE *
 __wt_update_obsolete_check(
-  WT_SESSION_IMPL *session, WT_PAGE *page, WT_UPDATE *upd, bool update_accounting)
+  WT_SESSION_IMPL *session, WT_PAGE *page, WT_UPDATE *upd, bool update_accounting, u_int *upd_count)
```
Comment (keithbostic) [inline src/btree/row_modify.c:314]: I'd go with `uint32_t` instead of `u_int`, just out of sheer raving paranoia.

File: src/btree/row_modify.c (hunk 6 above)
Comment (keithbostic) [reply src/btree/row_modify.c:337]: I think we're counting updates that we're going to discard in this loop, that is, we could have a long list of updates that we set the `WT_UPDATE *` return and then discard, but still evict the page. Honestly, it's hard to imagine that would make a difference, but thought I'd point it out.

File: src/include/serial_inline.h (hunk 4 above)
Comment (keithbostic) [reply src/include/serial_inline.h:308]: I'd suggest doing the page-eviction check and schedule in `__wt_update_obsolete_check`, rather than in its caller. There's only one caller and it simplifies the code, we don't need to pass back a count.

**Review by keithbostic — APPROVED** [review]
lgtm, +1 on @sulabhM's comments.

File: src/include/serial_inline.h (hunk 4 above)
Comment (tammybailey) [reply src/include/serial_inline.h:308]: Please add a comment as to why we are checking against the value 1000.

File: dist/stat_data.py (hunk 1 above)
Comment (raviprakashgiri29) [reply dist/stat_data.py]: Fixed

File: src/btree/row_modify.c (hunk 2 above)
Comment (raviprakashgiri29) [reply src/btree/row_modify.c:322]: I have changed the design a bit, now doing the page-eviction check and schedule in `__wt_update_obsolete_check` function.

File: src/btree/row_modify.c (hunk 7 above)
Comment (raviprakashgiri29) [reply src/btree/row_modify.c]: Removed this argument from the function `__wt_update_obsolete_check`

File: src/include/serial_inline.h (hunk 4 above)
Comment (raviprakashgiri29) [reply src/include/serial_inline.h]: > I'd suggest doing the page-eviction check and schedule in `__wt_update_obsolete_check`, rather than in its caller. There's only one caller and it simplifies the code, we don't need to pass back a count.

Fixed

Comment (raviprakashgiri29) [reply src/include/serial_inline.h]: > Instead of 1000 we can say WT_THOUSAND

Fixed

**Review by sulabhM — APPROVED** [review]
Much simpler now.

File: src/include/serial_inline.h (hunk 4 above)
Comment (raviprakashgiri29) [reply src/include/serial_inline.h]: @tammybailey Updated the comments, Can you please review the changes now?

File: src/btree/row_modify.c [hunk 8]
```diff
@@ -372,9 +375,21 @@ __wt_update_obsolete_check(
             if (size != 0)
                 __wt_cache_page_inmem_decr(session, page, size);
         }
-        return (next);
     }
 
+    /*
+     * Force evict a page when there are more than WT_THOUSAND updates to a single item. WT_THOUSAND
```
Comment (tammybailey) [inline src/btree/row_modify.c:381]: A few grammatical issues in the second sentence, maybe reword as:

"Increasing the minSnapshotHistoryWindowInSeconds to 300 introduced a performance regression in which the average number of updates on a single item was approximately 1000 in write-heavy workloads. This is why we use WT_THOUSAND here."

**Review by tammybailey — APPROVED** [review]
A few small grammatical issues in the comment and a suggested rewording.
lgtm otherwise.

File: src/btree/row_modify.c (hunk 8 above)
Comment (raviprakashgiri29) [reply src/btree/row_modify.c]: Fixed
